pkg/serverless/trace/inferredspan/span_enrichment.go

```diff
@@ -217,11 +220,125 @@ func (inferredSpan *InferredSpan) EnrichInferredSpanWithS3Event(eventPayload eve
+	TraceID          string `json:"x-datadog-trace-id"`
+	ParentID         string `json:"x-datadog-parent-id"`
+	SamplingPriority string `json:"x-datadog-sampling-priority"`
+	Sampled          string `json:"x-datadog-sampled"`
+}
+
+type customMessageAttributeStruct struct {
+	Type  string `json:"Type"`
+	Value string `json:"Value"`
+}
+type BodyStruct struct {
+	Message           string                                  `json:"Message"`
+	MessageAttributes map[string]customMessageAttributeStruct `json:"MessageAttributes"`
+}
+type CustomDatadogPayload struct {
+	Type        string `json:"Type"`
+	Value       string `json:"Value"`
+	DataType    string `json:"dataType"`
+	BinaryValue string `json:"binaryValue"`
+	StringValue string `json:"stringValue"`
+}
+
+func extractContextFromSNSSQSEvent(firstRecord events.SQSMessage) (string, string, string, error) {
+	var bodyStruct BodyStruct
+	if err := json.Unmarshal([]byte(firstRecord.Body), &bodyStruct); err != nil {
+		log.Errorf("Failed to parse Body field")
+		return "", "", "", err
+	}
+	var ddCustomPayloadValue, ok = bodyStruct.MessageAttributes["_datadog"]
+	if !ok {
+		log.Errorf("Datadog Lambda Go only supports extracting trace context from String or Binary SQS/SNS message attributes")
+		return "", "", "", errors.New("no Datadog trace context found")
+	}
+
+	var traceData TraceHeader
+	if ddCustomPayloadValue.Type == "Binary" {
+		decodedBinary, err := base64.StdEncoding.DecodeString(string(ddCustomPayloadValue.Value))
+		if err != nil {
+			log.Errorf("Failed to decode datadog binary payload: %s", err)
+			return "", "", "", err
+		}
+		if err := json.Unmarshal(decodedBinary, &traceData); err != nil {
+			log.Errorf("Failed to unmarshal datadog binary payload: %s", err)
+			return "", "", "", err
+		}
+	} else {
+		return "", "", "", errors.New("unsupported DataType in _datadog payload")
+	}
+
+	return traceData.TraceID, traceData.ParentID, traceData.SamplingPriority, nil
+}
+
+func extractContextFromPureSqsEvent(ddPayloadValue events.SQSMessageAttribute) (string, string, string, error) {
+	var traceData TraceHeader
+	if ddPayloadValue.DataType == "String" {
+		if err := json.Unmarshal([]byte(*ddPayloadValue.StringValue), &traceData); err != nil {
+			log.Errorf("Failed to unmarshal datadog string payload: %s", err)
+			return "", "", "", err
+		}
+	} else {
+		return "", "", "", errors.New("unsupported DataType in _datadog payload")
+	}
+
+	return traceData.TraceID, traceData.ParentID, traceData.SamplingPriority, nil
+}
+
 // EnrichInferredSpanWithSQSEvent uses the parsed event
 // payload to enrich the current inferred span. It applies a
 // specific set of data to the span expected from an SQS event.
 func (inferredSpan *InferredSpan) EnrichInferredSpanWithSQSEvent(eventPayload events.SQSEvent) {
+	var traceID, parentID, samplingPriority string
+	var err error
 	eventRecord := eventPayload.Records[0]
+	if ddMessageAttribute, ok := eventRecord.MessageAttributes["_datadog"]; ok {
```

Comment:
- Can we change the logic her so that we ONLY parseint/float when it successfully extracted the context? And only set inferredSpan.Span attributes when it successfully parsed it?
- But first, why do we need to parse them and not using string?
- The errs and logged and and passed along from the extract functions. Can we simplify this by only handle it in one place?

Reply:
> But first, why do we need to parse them and not using string?
TraceID, SpanID... [is expected to be a uint64](https://github.com/DataDog/datadog-agent/blob/4cdcc178d0263069f4a82336c079bbce7523291d/pkg/trace/pb/span.pb.go#L39)

Good thinking, I reformatted the logic to only handle the errors in the EnrichInferredSpanWithSQSEvent function additionally I added a conditional to adding the trace context to the inferred span instead of returning out prematurely